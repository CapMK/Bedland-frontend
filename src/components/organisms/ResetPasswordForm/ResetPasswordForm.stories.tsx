import type { Meta, StoryObj } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';
import GlobalStyle from 'theme/globalStyles';
import ThemeProvider from 'theme/ThemeContext';
import ResetPasswordForm from './ResetPasswordForm';

// docs: https://storybook.js.org/docs/7.0/react/writing-stories/introduction
const meta = {
  title: 'Organisms/ResetPasswordForm',
  component: ResetPasswordForm,
  tags: ['autodocs'],

  decorators: [
    (Story) => (
      <ThemeProvider>
        <GlobalStyle />
        <BrowserRouter>
          <div style={{ width: '25.625rem' }}>{Story()}</div>
        </BrowserRouter>
      </ThemeProvider>
    ),
  ],
} satisfies Meta<typeof ResetPasswordForm>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
