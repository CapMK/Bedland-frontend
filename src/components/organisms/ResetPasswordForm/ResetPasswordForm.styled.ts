import styled from 'styled-components';

export const ResetPasswordFormContainer = styled.form`
  display: flex;
  flex-direction: column;
  gap: 1rem;
  padding: 0 2.84375rem 0rem 2.84375rem;
  width: 100%;
`;

export default ResetPasswordFormContainer;
