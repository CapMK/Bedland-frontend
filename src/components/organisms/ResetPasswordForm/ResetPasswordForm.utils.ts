import ERROR_MESSAGES from 'utils/messages';

export const handleErrorMessage = (whichErrorMessageCondition: boolean): string =>
  whichErrorMessageCondition
    ? ERROR_MESSAGES.resetPasswordPage.notWithPasswordsRules
    : ERROR_MESSAGES.resetPasswordPage.passwordsDoesNotMatchUp;

export const watcher = (password: string, confirmPassword?: string) => {
  let passwordsMatchUp = false;

  if (confirmPassword !== undefined) {
    if (confirmPassword === password) passwordsMatchUp = true;
    else passwordsMatchUp = false;
  }
  return passwordsMatchUp;
};
