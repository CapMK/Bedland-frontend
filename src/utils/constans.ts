export const HEADER_HEIGHT = 70;
export const FOOTER_HEIGHT = 52;
export const LOGIN_INPUT_WIDTH = 19.875;

// TODO
// BASE_URL points to json server
// Replace with actual API url when BE is ready
export const BASE_URL = 'http://localhost:3001';
