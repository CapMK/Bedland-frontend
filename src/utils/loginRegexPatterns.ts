import { InputVariant, TRegister } from 'components/atoms/Input/Input.types';
import { Getters } from 'types/shared.types';

type LoginRegexProps = Getters<InputVariant, TRegister>;

const loginRegexPattern: LoginRegexProps = {
  login: {
    pattern: /^[a-zA-Z0-9]+$/,
    required: true,
    maxLength: 36,
    minLength: 8,
  },
  password: {
    pattern: /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[!@#$%^&*(),<.>/?;:'"]).{8,24}$/,
    required: true,
    maxLength: 36,
    minLength: 8,
  },
  confirmPassword: {
    required: true,
  },
};

export default loginRegexPattern;
// ^(?=.*[A-Za-z])(?=.*\d)(?=.*[@$!%*#?&])[A-Za-z\d@$!%*#?&]{8,}$ <- regex for password that should work.
// /^(?=.*[A-Za-z])(?=.*\d)[A-Za-z\d]{8,}$/,
