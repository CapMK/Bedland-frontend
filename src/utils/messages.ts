const ERROR_MESSAGES = {
  resetPasswordPage: {
    notWithPasswordsRules:
      'Password is incorrect, please try again after reviewing the rules above',

    passwordsDoesNotMatchUp: 'Passwords do not match to each other',
  },
  loginPage: {
    errorMessage: 'Your login or password is incorrect.',
  },
};

export default ERROR_MESSAGES;
