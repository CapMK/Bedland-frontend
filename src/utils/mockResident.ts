const mockResidentList = [
  {
    id: 1,
    version: 0,
    createDate: '2023-02-14T00:00:00',
    updateDate: '2023-02-14T00:00:00',
    login: 'iSutherland',
    password: 'IsABEl',
    name: 'Isabel',
    lastName: 'Sutherland',
    email: 'isabel.sutherland@email.com',
    phoneNumber: '123456789',
    flatId: 1,
    owner: true,
  },
  {
    id: 2,
    version: 0,
    createDate: '2023-02-14T00:00:00',
    updateDate: '2023-02-14T00:00:00',
    login: 'clWeiss',
    password: 'ClWeiss',
    name: 'Claude',
    lastName: 'Weiss',
    email: 'Claude.Weiss@email.com',
    phoneNumber: '123456789',
    flatId: 2,
    owner: false,
  },
];

export default mockResidentList;
