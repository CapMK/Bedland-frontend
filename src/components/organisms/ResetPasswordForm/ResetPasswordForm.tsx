import { useForm, SubmitHandler, FieldValues } from 'react-hook-form';
import ErrorMessage from 'components/atoms/ErrorMessage/ErrorMessage';
import { useNavigate } from 'react-router-dom';
import InputWithLabel from 'components/molecules/InputWithLabel/InputWithLabel';
import Button from 'components/atoms/Button/Button';
import * as Styled from 'components/atoms/ErrorMessage/ErrorMessage.styled';
import { RegisterOptions } from 'components/atoms/Input/Input.types';
import loginRegexPattern from 'utils/loginRegexPatterns';
import { handleErrorMessage, watcher } from './ResetPasswordForm.utils';
import { ResetPasswordFormContainer } from './ResetPasswordForm.styled';

const ResetPasswordForm = () => {
  const navigate = useNavigate();
  const {
    register,
    handleSubmit,
    watch,
    formState: { errors },
  } = useForm({ mode: 'onChange' });

  const enteredPasswordWatch = watch('password', '');

  const errorConditionWrongRules = Boolean(errors.password || errors.confirmPassword);

  let errorConditionIfPasswordsMatchUp = false;

  const whichErrorMessageCondition = Boolean(
    errors.password || (errorConditionIfPasswordsMatchUp && errors.password),
  );

  const resetPasswordRegOptions: RegisterOptions = {
    validate: (value) => {
      errorConditionIfPasswordsMatchUp = watcher(enteredPasswordWatch, value);
      return errorConditionIfPasswordsMatchUp;
    },
  };

  const onSubmit: SubmitHandler<FieldValues> = (data: FieldValues) => {
    // eslint-disable-next-line no-console
    console.log(data); // logic with fetch to the server should be here
    navigate('/');
  };

  return (
    <ResetPasswordFormContainer onSubmit={handleSubmit(onSubmit)}>
      <InputWithLabel
        label='Your new password'
        input='password'
        placeholder='Enter your new password'
        type='password'
        register={register}
        regexPattern={loginRegexPattern.password}
      />

      <InputWithLabel
        label='Confirm the password'
        input='confirmPassword'
        placeholder='Confirm the password'
        type='password'
        register={register}
        registerOptions={resetPasswordRegOptions}
        regexPattern={loginRegexPattern.password}
      />

      <Styled.ErrorMsg>
        {errorConditionWrongRules ? (
          <ErrorMessage>{handleErrorMessage(whichErrorMessageCondition)}</ErrorMessage>
        ) : null}
      </Styled.ErrorMsg>

      <Button type='submit' variant='text'>
        Reset my password
      </Button>
    </ResetPasswordFormContainer>
  );
};

export default ResetPasswordForm;
