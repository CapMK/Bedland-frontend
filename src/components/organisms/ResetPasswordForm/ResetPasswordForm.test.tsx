import { act, fireEvent, screen } from '@testing-library/react';
import { renderWithProviders } from 'utils/test.utils';
import ERROR_MESSAGES from 'utils/messages';
import { vi } from 'vitest';
import ResetPasswordForm from './ResetPasswordForm';
import { handleErrorMessage, watcher } from './ResetPasswordForm.utils';

const mockUseNavigate = vi.fn();

vi.mock('react-router-dom', async () => ({
  ...((await vi.importActual('react-router-dom')) as any),
  useNavigate: () => mockUseNavigate,
}));

describe('ResetPassword Form', () => {
  it('renders correctly', () => {
    const snapshot = renderWithProviders(<ResetPasswordForm />, {});
    expect(snapshot).toMatchSnapshot();
  });

  it('The password is according to the rules and passwords match', async () => {
    renderWithProviders(<ResetPasswordForm />, {});
    let resetPasswordButton: Element | null = null;
    let inputPassword: Element | null = null;
    let inputConfirmPassword: Element | null = null;

    await act(async () => {
      resetPasswordButton = screen.getByRole('button');
      inputPassword = screen.getByPlaceholderText('Enter your new password');
      inputConfirmPassword = screen.getByPlaceholderText('Confirm the password');

      fireEvent.input(inputPassword, {
        target: {
          value: 'Aaaaaaaaa2.',
        },
      });

      fireEvent.input(inputConfirmPassword, {
        target: {
          value: 'Aaaaaaaaa2.',
        },
      });

      fireEvent.submit(resetPasswordButton as HTMLElement);
    });

    expect(mockUseNavigate).toHaveBeenCalled();
    expect(mockUseNavigate).toHaveBeenCalledTimes(1);
    expect(mockUseNavigate).toHaveBeenCalledWith('/');
  });

  it('The password is according to the rules and passwords dont match', async () => {
    renderWithProviders(<ResetPasswordForm />, {});
    let resetPasswordButton: Element | null = null;
    let inputPassword: Element | null = null;
    let inputConfirmPassword: Element | null = null;

    await act(async () => {
      resetPasswordButton = screen.getByRole('button');
      inputPassword = screen.getByPlaceholderText('Enter your new password');
      inputConfirmPassword = screen.getByPlaceholderText('Confirm the password');

      fireEvent.input(inputPassword, {
        target: {
          value: 'Aaaaaaaaa2.',
        },
      });

      fireEvent.input(inputConfirmPassword, {
        target: {
          value: 'Aaaaaaaaa2',
        },
      });

      fireEvent.submit(resetPasswordButton as HTMLElement);
    });

    expect(
      await screen.getByText(ERROR_MESSAGES.resetPasswordPage.passwordsDoesNotMatchUp),
    ).not.toBeNull();
  });

  it('The password is not according to the rules and passwords match ', async () => {
    renderWithProviders(<ResetPasswordForm />, {});
    let resetPasswordButton: Element | null = null;
    let inputPassword: Element | null = null;
    let inputConfirmPassword: Element | null = null;

    await act(async () => {
      resetPasswordButton = screen.getByRole('button');
      inputPassword = screen.getByPlaceholderText('Enter your new password');
      inputConfirmPassword = screen.getByPlaceholderText('Confirm the password');

      fireEvent.input(inputPassword, {
        target: {
          value: 'Aaaaaaaaa2',
        },
      });

      fireEvent.input(inputConfirmPassword, {
        target: {
          value: 'Aaaaaaaaa2',
        },
      });

      fireEvent.submit(resetPasswordButton as HTMLElement);
    });

    // expect(window.location.pathname).toBe('/');

    expect(
      await screen.getByText(ERROR_MESSAGES.resetPasswordPage.notWithPasswordsRules),
    ).not.toBeNull();
  });
  it('The password is not according to the rules and passwords dont match ', async () => {
    renderWithProviders(<ResetPasswordForm />, {});
    let resetPasswordButton: Element | null = null;
    let inputPassword: Element | null = null;
    let inputConfirmPassword: Element | null = null;

    await act(async () => {
      resetPasswordButton = screen.getByRole('button');
      inputPassword = screen.getByPlaceholderText('Enter your new password');
      inputConfirmPassword = screen.getByPlaceholderText('Confirm the password');

      fireEvent.input(inputPassword, {
        target: {
          value: 'Aaaaaaaaa2',
        },
      });

      fireEvent.input(inputConfirmPassword, {
        target: {
          value: 'Aaaaaaaaa.2',
        },
      });

      fireEvent.submit(resetPasswordButton as HTMLElement);
    });

    // expect(window.location.pathname).toBe('/');

    expect(
      await screen.getByText(ERROR_MESSAGES.resetPasswordPage.notWithPasswordsRules),
    ).not.toBeNull();
  });
});
describe('ResetPassword Utils', () => {
  it('should return the correct error message when passwords do not match up', () => {
    const password = 'password123';
    const confirmPassword = 'password1234';

    const result = handleErrorMessage(watcher(password, confirmPassword));

    expect(result).toBe(ERROR_MESSAGES.resetPasswordPage.passwordsDoesNotMatchUp);
  });

  it('should return the correct error message when passwords do not meet requirements', () => {
    const password = 'pass';
    const confirmPassword = 'pass';

    const result = handleErrorMessage(watcher(password, confirmPassword));

    expect(result).toBe(ERROR_MESSAGES.resetPasswordPage.notWithPasswordsRules);
  });
});
