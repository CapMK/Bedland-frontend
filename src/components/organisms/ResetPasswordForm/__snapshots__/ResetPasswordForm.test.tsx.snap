// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResetPassword Form > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": .c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.875rem;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  width: 100%;
}

.c3 {
  border: 1px solid #919AFF;
  border-radius: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.875rem;
  padding: 0.625rem 1rem 0.625rem 1rem;
  width: 19.875rem;
}

.c3:disabled,
.c3:hover,
.c3:active,
.c3:focus {
  outline: none;
}

.c2 {
  color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 16;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.6875rem;
  width: 19.875rem;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border: 1px solid transparent;
  border-radius: 0.5rem;
  color: #FF7C33;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.5625rem 1rem;
  white-space: nowrap;
  width: 100%;
  font-family: Source Sans Pro,sans-serif;
  font-size: 1rem;
  font-weight: 600;
  line-height: 1.4;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

.c5:hover {
  background-color: #FF7C330D;
  border: 1px solid transparent;
}

.c6 {
  background-color: transparent;
  border: 1.75px solid transparent;
  color: #FF7C33;
  padding: undefinedrem 1rem;
}

.c6:hover {
  background-color: #FF7C330D;
  border: 1.75px solid transparent;
}

.c6:focus {
  background-color: #FF7C3326;
  border: 1.75px solid transparent;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1rem;
  padding: 0 2.84375rem 0rem 2.84375rem;
  width: 100%;
}

<body>
    <div>
      <form
        class="c0"
      >
        <div
          class="c1"
        >
          <label
            class="c2"
            for="password"
          >
            Your new password
          </label>
          <input
            class="c3"
            data-testid="password"
            id="password"
            name="password"
            placeholder="Enter your new password"
            type="password"
          />
        </div>
        <div
          class="c1"
        >
          <label
            class="c2"
            for="confirmPassword"
          >
            Confirm the password
          </label>
          <input
            class="c3"
            data-testid="confirmPassword"
            id="confirmPassword"
            name="confirmPassword"
            placeholder="Confirm the password"
            type="password"
          />
        </div>
        <span
          class="c4"
        />
        <button
          class="c5 c6"
          type="submit"
        >
          Reset my password
        </button>
      </form>
      <div
        data-testid="location-display"
      >
        /
      </div>
    </div>
  </body>,
  "container": .c4 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.875rem;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  width: 100%;
}

.c3 {
  border: 1px solid #919AFF;
  border-radius: 0.5rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.875rem;
  padding: 0.625rem 1rem 0.625rem 1rem;
  width: 19.875rem;
}

.c3:disabled,
.c3:hover,
.c3:active,
.c3:focus {
  outline: none;
}

.c2 {
  color: white;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 16;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 0.6875rem;
  width: 19.875rem;
}

.c5 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: transparent;
  border: 1px solid transparent;
  border-radius: 0.5rem;
  color: #FF7C33;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 10px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  padding: 0.5625rem 1rem;
  white-space: nowrap;
  width: 100%;
  font-family: Source Sans Pro,sans-serif;
  font-size: 1rem;
  font-weight: 600;
  line-height: 1.4;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

.c5:hover {
  background-color: #FF7C330D;
  border: 1px solid transparent;
}

.c6 {
  background-color: transparent;
  border: 1.75px solid transparent;
  color: #FF7C33;
  padding: undefinedrem 1rem;
}

.c6:hover {
  background-color: #FF7C330D;
  border: 1.75px solid transparent;
}

.c6:focus {
  background-color: #FF7C3326;
  border: 1.75px solid transparent;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 1rem;
  padding: 0 2.84375rem 0rem 2.84375rem;
  width: 100%;
}

<div>
    <form
      class="c0"
    >
      <div
        class="c1"
      >
        <label
          class="c2"
          for="password"
        >
          Your new password
        </label>
        <input
          class="c3"
          data-testid="password"
          id="password"
          name="password"
          placeholder="Enter your new password"
          type="password"
        />
      </div>
      <div
        class="c1"
      >
        <label
          class="c2"
          for="confirmPassword"
        >
          Confirm the password
        </label>
        <input
          class="c3"
          data-testid="confirmPassword"
          id="confirmPassword"
          name="confirmPassword"
          placeholder="Confirm the password"
          type="password"
        />
      </div>
      <span
        class="c4"
      />
      <button
        class="c5 c6"
        type="submit"
      >
        Reset my password
      </button>
    </form>
    <div
      data-testid="location-display"
    >
      /
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
